src/main/groovy/com/google/protobuf/gradle/ToolsLocator.groovy

```diff
@@ -108,4 +108,13 @@ class ToolsLocator {
       }
     }
   }
+
+  static List<String> artifactParts(String artifactCoordinate) {
+    final validName = '[+-.0-9A-Z_a-z]+'
+
+    ((artifactCoordinate =~ /(${validName}):(${validName}):(${validName})(:${validName})?(@${validName})?/)[0]
```

Comment:
I was thinking something along the lines of this, which passes your unit tests, but may trip some code narc lint rules when written this particular way:
```
  static List<String> artifactParts(String artifactCoordinate) {
    def (name, extension) = artifactCoordinate.tokenize("@")
    def (groupId, artifact, version, classifier) = name.tokenize(":")
    return [groupId, artifact, version, classifier, extension]
  }
```